File: client/lib/core/sync/sync_dto.dart
/// Mirrors backend/src/Purch.Application/Sync/SyncDto.cs — keep in sync with
/// that file's shape.
class SyncItemRequest {
  SyncItemRequest({
    required this.idempotencyKey,
    required this.entityType,
    required this.entityId,
    required this.clientTimestamp,
    required this.payloadJson,
  });

  final String idempotencyKey;
  final String entityType;
  final String entityId;
  final DateTime clientTimestamp;
  final String payloadJson;

  Map<String, dynamic> toJson() => {
    'idempotencyKey': idempotencyKey,
    'entityType': entityType,
    'entityId': entityId,
    'clientTimestamp': clientTimestamp.toUtc().toIso8601String(),
    'payloadJson': payloadJson,
  };
}

enum SyncItemStatus { applied, alreadySynced, conflictFlagged }

SyncItemStatus _syncItemStatusFromJson(String value) {
  switch (value) {
    case 'Applied':
      return SyncItemStatus.applied;
    case 'AlreadySynced':
      return SyncItemStatus.alreadySynced;
    case 'ConflictFlagged':
      return SyncItemStatus.conflictFlagged;
    default:
      throw ArgumentError('Unknown SyncItemStatus: $value');
  }
}

class SyncItemResult {
  SyncItemResult({
    required this.idempotencyKey,
    required this.status,
    required this.message,
  });

  factory SyncItemResult.fromJson(Map<String, dynamic> json) {
    return SyncItemResult(
      idempotencyKey: json['idempotencyKey'] as String,
      status: _syncItemStatusFromJson(json['status'] as String),
      message: json['message'] as String,
    );
  }

  final String idempotencyKey;
  final SyncItemStatus status;
  final String message;
}

class SyncBatchResult {
  SyncBatchResult({required this.results});

  factory SyncBatchResult.fromJson(Map<String, dynamic> json) {
    return SyncBatchResult(
      results:
          (json['results'] as List<dynamic>)
              .cast<Map<String, dynamic>>()
              .map(SyncItemResult.fromJson)
              .toList(),
    );
  }

  final List<SyncItemResult> results;
}

class FlaggedSyncRecord {
  FlaggedSyncRecord({
    required this.id,
    required this.deviceId,
    required this.entityType,
    required this.entityId,
    required this.clientTimestamp,
    required this.reviewedAt,
  });

  factory FlaggedSyncRecord.fromJson(Map<String, dynamic> json) {
    return FlaggedSyncRecord(
      id: json['id'] as String,
      deviceId: json['deviceId'] as String,
      entityType: json['entityType'] as String,
      entityId: json['entityId'] as String,
      clientTimestamp: DateTime.parse(json['clientTimestamp'] as String),
      reviewedAt:
          json['reviewedAt'] == null
              ? null
              : DateTime.parse(json['reviewedAt'] as String),
    );
  }

  final String id;
  final String deviceId;
  final String entityType;
  final String entityId;
  final DateTime clientTimestamp;
  final DateTime? reviewedAt;
}

File: backend/src/Purch.Application/Sync/SyncDto.cs
namespace Purch.Application.Sync;

/// <summary>One queued offline write, as the client's local
/// pending_sync_queue table stores it — see docs/ARCHITECTURE.md §5.</summary>
public sealed record SyncItemRequest(
    string IdempotencyKey,
    string EntityType,
    Guid EntityId,
    DateTimeOffset ClientTimestamp,
    string PayloadJson);

public sealed record SyncBatchRequest(IReadOnlyList<SyncItemRequest> Items);

public enum SyncItemStatus
{
    /// <summary>Recorded as the current winner for its entity.</summary>
    Applied,

    /// <summary>This exact IdempotencyKey was already recorded — safe to
    /// discard from the local queue, nothing new happened.</summary>
    AlreadySynced,

    /// <summary>A different device already holds (or now holds) the winning
    /// state for this entity and this item lost the later-timestamp
    /// comparison — flagged for manual review, not applied, not dropped.</summary>
    ConflictFlagged,
}

public sealed record SyncItemResultDto(string IdempotencyKey, SyncItemStatus Status, string Message);

public sealed record SyncBatchResultDto(IReadOnlyList<SyncItemResultDto> Results);

/// <summary>A record flagged by the conflict rule, for F-series manual review.</summary>
public sealed record FlaggedSyncRecordDto(
    Guid Id,
    Guid DeviceId,
    string EntityType,
    Guid EntityId,
    DateTimeOffset ClientTimestamp,
    DateTimeOffset? ReviewedAt);
